python/tests/catalog/test_base.py

```diff
@@ -0,0 +1,444 @@
+        location: Optional[str] = None,
+        partition_spec: Optional[PartitionSpec] = None,
+        properties: Optional[Dict[str, str]] = None,
+    ) -> Table:
+
+        try:
+            table = self.__tables.pop((namespace, name))
+        except KeyError:
+            raise TableNotFoundError(f"Table {name} not found in the catalog")
+
+        replaced_table = InMemoryTable(
+            namespace=namespace if namespace else table.namespace,
+            name=name if name else table.name,
+            schema=schema if schema else table.schema,
+            location=location if location else table.location,
+            partition_spec=partition_spec if partition_spec else table.partition_spec,
+            properties={**table.properties, **properties},
+        )
+        self.__tables[(replaced_table.namespace, replaced_table.name)] = replaced_table
+        return replaced_table
+
+    def create_namespace(self, namespace: Tuple[str, ...], properties: Optional[Dict[str, str]] = None) -> None:
+        if namespace in self.__namespaces:
+            raise AlreadyExistsError(f"Namespace {namespace} already exists")
+        else:
+            self.__namespaces[namespace] = properties if properties else {}
+
+    def drop_namespace(self, namespace: Tuple[str, ...]) -> None:
+        if [table_name_tuple for table_name_tuple in self.__tables.keys() if namespace in table_name_tuple]:
+            raise NamespaceNotEmptyError(f"Namespace {namespace} not empty")
+        try:
+            self.__namespaces.pop(namespace)
+        except KeyError:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+    def list_tables(self, namespace: Optional[Tuple[str, ...]] = None) -> List[Tuple[Tuple[str, ...], str]]:
+        if namespace:
+            list_tables = [table_name_tuple for table_name_tuple in self.__tables.keys() if namespace in table_name_tuple]
+        else:
+            list_tables = list(self.__tables.keys())
+
+        # Casting to make mypy happy
+        return cast(List[Tuple[Tuple[str, ...], str]], list_tables)
+
+    def list_namespaces(self) -> List[Tuple[str, ...]]:
+        return list(self.__namespaces.keys())
+
+    def load_namespace_metadata(self, namespace: Tuple[str, ...]) -> Dict[str, str]:
+        try:
+            return self.__namespaces[namespace]
+        except KeyError:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+    def set_namespace_metadata(self, namespace: Tuple[str, ...], metadata: Dict[str, str]) -> None:
+        if namespace in self.__namespaces:
+            self.__namespaces[namespace] = metadata
+        else:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+
+@pytest.fixture
+def catalog() -> InMemoryCatalog:
+    return InMemoryCatalog("test.in.memory.catalog", {"test.key": "test.value"})
+
+
+@pytest.fixture
+def table_spec() -> InMemoryTable:
```

Comment:
Since we're exploding the arguments to the catalog methods now, do we need this fixture or the `InMemoryTable` class above? At first glance it may imply that catalog functionality requires you provide `Table` instances. Maybe we can just use a file-level constant that's just a dictionary.
```py
EXAMPLE_TABLE1 = {
  "namespace": ("com", "organization", "department"),
  "name": "my_table",
  "schema": Schema(schema_id=1),
  "location": "protocol://some/location",
  "partition_spec": PartitionSpec(),
  "properties": {"key1": "value1", "key2": "value2"}
}
```

Reply:
As suggested above, I've removed the table_spec fixture and just using constants for these test values in my test now.

I've removed the `InMemoryTable` implementation of `Table(ABC)` and trimmed down the InMemoryCatalog implementation to not use it and just use a bland `Table` instance. The only reason `InMemoryTable` existed was because I didn't want to add any table attributes to `Table(ABC)` in this PR to keep it as slim as possible. I had to sacrifice some of the `InMemoryCatalog` functionality around `rename_table()` and `replace_table()` but it still does a decent job of testing the base catalog interface and we can always add that back once we have a base `Table` implementation with all the attributes.